File: core/interrelations.py
"""Interrelation matrix: manages and applies inter-aspect influence propagation."""

import numpy as np


class InterrelationMatrix:
    """Aspect coupling matrix supporting symmetric pairs, asymmetric weighted
    edges, and inhibitory connections."""

    def __init__(self):
        self._matrix: np.ndarray | None = None
        self._aspects: list[str] = []

    def build(self, aspects: list[str], relationships: list, strength: float = 0.005):
        """Construct the matrix from relationship definitions.

        Relationships can be:
          - (a, b)       — symmetric pair at `strength`
          - (a, b, w)    — directed edge a→b with weight w
        """
        n = len(aspects)
        self._aspects = list(aspects)
        self._matrix = np.identity(n)
        idx = {a: i for i, a in enumerate(aspects)}

        for rel in relationships:
            if len(rel) == 2:
                a1, a2 = rel
                w = strength
                if a1 in idx and a2 in idx:
                    self._matrix[idx[a1], idx[a2]] = w
                    self._matrix[idx[a2], idx[a1]] = w
            elif len(rel) == 3:
                a1, a2, w = rel
                if a1 in idx and a2 in idx:
                    self._matrix[idx[a1], idx[a2]] = w

    def propagate(self, weights: list[float]) -> list[float]:
        """Apply the interrelation matrix to a weight vector."""
        if self._matrix is None:
            return list(weights)
        w = np.array(weights)
        return (self._matrix @ w).tolist()

    @property
    def matrix(self) -> np.ndarray:
        return self._matrix if self._matrix is not None else np.array([])

File: native/src/interrelations.rs
//! Interrelation matrix: manages and applies inter-aspect influence propagation.
//!
//! Mirrors the Python `InterrelationMatrix` in core/interrelations.py exactly.
//! Triples `(src, dst, weight)` set directed edges; the matrix is initialised
//! as the identity so unspecified pairs leave weights unchanged.

pub struct InterrelationMatrix {
    data: Vec<f64>, // row-major n×n
    n: usize,
}

impl InterrelationMatrix {
    /// Create an n×n identity matrix (no coupling yet).
    pub fn new(n: usize) -> Self {
        let mut data = vec![0.0_f64; n * n];
        for i in 0..n {
            data[i * n + i] = 1.0;
        }
        Self { data, n }
    }

    /// Rebuild from (src, dst, weight) triples.
    ///
    /// Resets to identity first, then sets `matrix[src][dst] = weight` for
    /// each triple.  Symmetric pairs must be passed as two separate triples
    /// (the Python layer handles that before calling here).
    pub fn build(&mut self, triples: &[(usize, usize, f64)]) {
        // Reset to identity.
        self.data.iter_mut().for_each(|x| *x = 0.0);
        for i in 0..self.n {
            self.data[i * self.n + i] = 1.0;
        }
        // Apply edges.
        for &(src, dst, w) in triples {
            self.data[src * self.n + dst] = w;
        }
    }

    /// Matrix-vector multiply: `out = self.data @ weights`.
    ///
    /// `out[i] = Σ_j data[i*n + j] * weights[j]`
    ///
    /// Manual loop — for the 20-32 aspect sizes used here the BLAS call
    /// overhead exceeds the compute cost.
    pub fn propagate(&self, weights: &[f64], out: &mut [f64]) {
        let n = self.n;
        for (i, slot) in out.iter_mut().enumerate().take(n) {
            let row = &self.data[i * n..(i + 1) * n];
            let mut acc = 0.0_f64;
            for j in 0..n {
                acc += row[j] * weights[j];
            }
            *slot = acc;
        }
    }

    pub fn matrix(&self) -> &[f64] {
        &self.data
    }

    pub fn n(&self) -> usize {
        self.n
    }
}
